src/main/starlark/builtins_bzl/common/cc/experimental_cc_shared_library.bzl

```diff
@@ -72,16 +107,45 @@ def _separate_static_and_dynamic_link_libraries(
             break
 
         node = all_children[i]
-        node_label = str(node.label)
-
-        if node_label in seen_labels:
-            continue
-        seen_labels[node_label] = True
 
-        if node_label in can_be_linked_dynamically:
-            targets_to_be_linked_dynamically_set[node_label] = True
-        else:
-            targets_to_be_linked_statically_map[node_label] = node.linkable_more_than_once
+        must_add_children = False
+        # The seen count is used to track a programmatic error and fail if it
+        # happens.  Every value in node.owners presumably corresponds to a
+        # linker_input in the same exact target. Therefore if we have seen any
+        # of the owners already, then we must have also seen all the other
+        # owners in the same node. Viceversa when we haven't seen them yet. If
+        # both of these values are non-zero after the loop, the most likely
+        # reason would be a bug in the implementation. It could potentially be
+        # triggered by users if they use owner labels that do not keep most of
+        # the ctx.label.package and ctx.label.name which then clash with other
+        # target's owners (unlikely). For now though if the error is
+        # triggered, it's reasonable to require manual revision by
+        # the cc_shared_library implementation owners.
+        seen_count = 0
+        not_seen_count = 0
+        for owner in node.owners:
+            owner_str = str(owner)
+
+            if owner_str in seen_labels:
+                seen_count += 1
+                continue
+
+            not_seen_count += 1
+
+
+            seen_labels[owner_str] = True
+
+            if owner_str in can_be_linked_dynamically:
+                targets_to_be_linked_dynamically_set[owner_str] = True
+            else:
+                targets_to_be_linked_statically_map[owner_str] = node.linkable_more_than_once
+                must_add_children = True
+
+        if seen_count and not_seen_count:
```

Comment:
not_seen_count will always be more than 0 if node.owners is not empty. Basically you can get rid of not_seen_count variable and rewrite this if as:
```
if len(node.owners) > 0 and seen_count:
  # Do something
```

I do not think having not_seen_count as a separate variable makes sense since it is just a proxy check if node.owners is empty or not.

Also seen_count and not_seen_count(if you decide to keep it) can be boolean instead of int.

Reply:
I can change it to booleans but must keep the two variables  (discussed offline).